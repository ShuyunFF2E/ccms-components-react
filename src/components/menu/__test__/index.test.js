import React from 'react';
import { render, mount } from 'enzyme';

import Menu from '../index';

const menuSource = [
	{
		title: '北京',
		key: '/beijing',
		children: [
			{
				title: <span style={{ color: 'red', fontWeight: 'bold' }}>朝阳区</span>,
				key: '/beijing/chaoyang'
			},
			{
				title: '海淀区',
				key: '/beijing/haidian'
			}
		]
	},
	{
		title: '上海',
		key: '/shanghai',
		children: [
			{
				title: '浦东区',
				key: '/shanghai/pudong'
			},
			{
				title: '南京路',
				key: '#/shanghai/nanjinglu'
			}
		]
	}
];

describe('Menu', () => {
	it('renders correctly with composition', () => {
		const wrapper = render(
			<Menu>
				<Menu.MenuItem key="1">menuItem1</Menu.MenuItem>
				<Menu.SubMenu key="2" title="subMenu2">
					<Menu.MenuItem key="21">menuItem2</Menu.MenuItem>
					<Menu.MenuItem key="22">menuItem3</Menu.MenuItem>
				</Menu.SubMenu>
			</Menu>
		);
		expect(wrapper).toMatchSnapshot();
	});

	it('renders correctly with structual data', () => {
		const wrapper = render(<Menu source={menuSource}></Menu>);
		expect(wrapper).toMatchSnapshot();
	});

	it('openKeys works correctly', () => {
		const wrapper = mount(<Menu source={menuSource} openKeys={['/beijing']}></Menu>);

		// initial
		expect(
			wrapper
				.find('.group-sub')
				.at(0)
				.hasClass('expand')
		).toBeTruthy();
		expect(
			wrapper
				.find('.group-sub')
				.at(1)
				.hasClass('collapse')
		).toBeTruthy();

		// props change
		wrapper.setProps({ openKeys: ['/beijing', '/shanghai'] });

		expect(wrapper.state().openKeys).toEqual(['/beijing', '/shanghai']);

		expect(
			wrapper
				.find('.group-sub')
				.at(0)
				.hasClass('expand')
		).toBeTruthy();

		expect(
			wrapper
				.find('.group-sub')
				.at(1)
				.hasClass('expand')
		).toBeTruthy();
	});

	it('selectedKeys works correctly', () => {
		const wrapper = mount(<Menu source={menuSource} selectedKeys="/beijing/haidian"></Menu>);

		// initial
		expect(
			wrapper
				.find('.group-sub')
				.at(0)
				.find('.item')
				.at(1)
				.hasClass('active')
		).toBeTruthy();

		// props change
		wrapper.setProps({ selectedKeys: '/beijing/chaoyang' });
		expect(
			wrapper
				.find('.group-sub')
				.at(0)
				.find('.item')
				.at(0)
				.hasClass('active')
		).toBeTruthy();
	});

	it("parent/ancestor SubMenu could be opend, when selectedKeys startsWith parent SubMenu's key", () => {
		const wrapper = mount(
			<Menu selectedKeys="abcde">
				<Menu.SubMenu key="abc" title="第一层">
					<Menu.SubMenu key="abcd" title="第二层">
						<Menu.MenuItem key="abcde">最内层</Menu.MenuItem>
					</Menu.SubMenu>
				</Menu.SubMenu>
			</Menu>
		);
		expect(
			wrapper
				.find('.item')
				.at(0)
				.hasClass('active')
		).toBeTruthy();
		expect(
			wrapper
				.find('.group-sub')
				.at(0)
				.hasClass('expand')
		).toBeTruthy();
		expect(
			wrapper
				.find('.group-sub .group-sub')
				.at(0)
				.hasClass('expand')
		).toBeTruthy();
	});

	it('onItemClick works correctly', () => {
		const handleItemClick = jest.fn();
		const wrapper = mount(<Menu source={menuSource} onItemClick={handleItemClick}></Menu>);
		wrapper
			.find('.item')
			.at(0)
			.simulate('click');
		expect(handleItemClick).toBeCalledWith('/beijing/chaoyang', ['/beijing/chaoyang', '/beijing']);
	});

	it('onSubMenuToggle works correctly', () => {
		const handleSubMenuToggle = jest.fn();
		const wrapper = mount(<Menu source={menuSource} onSubMenuToggle={handleSubMenuToggle}></Menu>);
		const target = wrapper.find('.group-title').at(0);

		target.simulate('click');
		expect(handleSubMenuToggle).toBeCalledWith('/beijing', ['/beijing'], true);

		target.simulate('click');
		expect(handleSubMenuToggle).toBeCalledWith('/beijing', ['/beijing'], false);
	});

	it('header works correctly', () => {
		const handleClick = jest.fn();
		const wrapper = mount(
			<Menu
				source={menuSource}
				header={
					<span className="my-header" onClick={handleClick}>
						user header
					</span>
				}></Menu>
		);

		// display correctly
		const target = wrapper.find('.my-header').at(0);
		expect(target.hasClass('my-header')).toBeTruthy();
		expect(wrapper).toMatchSnapshot();

		// event
		target.simulate('click');
		expect(handleClick).toBeCalled();
	});

	it('type works correctly', () => {
		const wrapper = mount(<Menu source={menuSource} type="link"></Menu>);

		expect(
			wrapper
				.find('a>.item')
				.at(0)
				.hasClass('item')
		).toBeTruthy();

		wrapper.setProps({ type: 'common' });
		expect(wrapper.find('a>.item')).toEqual({});
	});
});
