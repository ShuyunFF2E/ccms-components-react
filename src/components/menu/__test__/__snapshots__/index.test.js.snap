// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu header works correctly 1`] = `
<Menu
  header={
    <span
      className="my-header"
      onClick={[MockFunction]}
    >
      user header
    </span>
  }
  indent={10}
  onItemClick={[Function]}
  onSubMenuToggle={[Function]}
  openKeys={Array []}
  selectedKeys={Array []}
  source={
    Array [
      Object {
        "children": Array [
          Object {
            "key": "/beijing/chaoyang",
            "title": <span
              style={
                Object {
                  "color": "red",
                  "fontWeight": "bold",
                }
              }
            >
              朝阳区
            </span>,
          },
          Object {
            "key": "/beijing/haidian",
            "title": "海淀区",
          },
        ],
        "key": "/beijing",
        "title": "北京",
      },
      Object {
        "children": Array [
          Object {
            "key": "/shanghai/pudong",
            "title": "浦东区",
          },
          Object {
            "key": "/shanghai/nanjinglu",
            "title": "南京路",
          },
        ],
        "key": "/shanghai",
        "title": "上海",
      },
    ]
  }
  style={Object {}}
  type="common"
>
  <section
    className="cloud-menu-container"
    style={Object {}}
  >
    <span
      className="my-header"
      onClick={[MockFunction]}
    >
      user header
    </span>
    <ul
      className="list"
    >
      <SubMenu
        indent={10}
        internalKey="/beijing"
        key="/beijing"
        onMenuToggle={[Function]}
        opened={false}
        path={
          Array [
            "/beijing",
          ]
        }
        title="北京"
      >
        <li>
          <div
            className="group-title"
            onClick={[Function]}
            style={
              Object {
                "textIndent": 20,
              }
            }
          >
            北京
            <Icon
              className="indicator"
              style={Object {}}
              type="down"
            >
              <i
                className="indicator cloud-icon cloud-icon-down"
                style={Object {}}
              />
            </Icon>
          </div>
          <ul
            className="group-sub collapse"
          >
            <MenuItem
              indent={10}
              internalKey="/beijing/chaoyang"
              key="/beijing/chaoyang"
              onClick={[Function]}
              path={
                Array [
                  "/beijing/chaoyang",
                  "/beijing",
                ]
              }
              selected={false}
              type="common"
            >
              <li
                className="item"
                onClick={[Function]}
                role="menuitem"
                style={
                  Object {
                    "textIndent": 30,
                  }
                }
              >
                <span
                  style={
                    Object {
                      "color": "red",
                      "fontWeight": "bold",
                    }
                  }
                >
                  朝阳区
                </span>
              </li>
            </MenuItem>
            <MenuItem
              indent={10}
              internalKey="/beijing/haidian"
              key="/beijing/haidian"
              onClick={[Function]}
              path={
                Array [
                  "/beijing/haidian",
                  "/beijing",
                ]
              }
              selected={false}
              type="common"
            >
              <li
                className="item"
                onClick={[Function]}
                role="menuitem"
                style={
                  Object {
                    "textIndent": 30,
                  }
                }
              >
                海淀区
              </li>
            </MenuItem>
          </ul>
        </li>
      </SubMenu>
      <SubMenu
        indent={10}
        internalKey="/shanghai"
        key="/shanghai"
        onMenuToggle={[Function]}
        opened={false}
        path={
          Array [
            "/shanghai",
          ]
        }
        title="上海"
      >
        <li>
          <div
            className="group-title"
            onClick={[Function]}
            style={
              Object {
                "textIndent": 20,
              }
            }
          >
            上海
            <Icon
              className="indicator"
              style={Object {}}
              type="down"
            >
              <i
                className="indicator cloud-icon cloud-icon-down"
                style={Object {}}
              />
            </Icon>
          </div>
          <ul
            className="group-sub collapse"
          >
            <MenuItem
              indent={10}
              internalKey="/shanghai/pudong"
              key="/shanghai/pudong"
              onClick={[Function]}
              path={
                Array [
                  "/shanghai/pudong",
                  "/shanghai",
                ]
              }
              selected={false}
              type="common"
            >
              <li
                className="item"
                onClick={[Function]}
                role="menuitem"
                style={
                  Object {
                    "textIndent": 30,
                  }
                }
              >
                浦东区
              </li>
            </MenuItem>
            <MenuItem
              indent={10}
              internalKey="/shanghai/nanjinglu"
              key="/shanghai/nanjinglu"
              onClick={[Function]}
              path={
                Array [
                  "/shanghai/nanjinglu",
                  "/shanghai",
                ]
              }
              selected={false}
              type="common"
            >
              <li
                className="item"
                onClick={[Function]}
                role="menuitem"
                style={
                  Object {
                    "textIndent": 30,
                  }
                }
              >
                南京路
              </li>
            </MenuItem>
          </ul>
        </li>
      </SubMenu>
    </ul>
  </section>
</Menu>
`;

exports[`Menu renders correctly with composition 1`] = `
<section
  class="cloud-menu-container"
>
  <ul
    class="list"
  >
    <li
      class="item"
      role="menuitem"
      style="text-indent:20px"
    >
      menuItem1
    </li>
    <li>
      <div
        class="group-title"
        style="text-indent:20px"
      >
        subMenu2
        <i
          class="indicator cloud-icon cloud-icon-down"
        />
      </div>
      <ul
        class="group-sub collapse"
      >
        <li
          class="item"
          role="menuitem"
          style="text-indent:30px"
        >
          menuItem2
        </li>
        <li
          class="item"
          role="menuitem"
          style="text-indent:30px"
        >
          menuItem3
        </li>
      </ul>
    </li>
  </ul>
</section>
`;

exports[`Menu renders correctly with structual data 1`] = `
<section
  class="cloud-menu-container"
>
  <ul
    class="list"
  >
    <li>
      <div
        class="group-title"
        style="text-indent:20px"
      >
        北京
        <i
          class="indicator cloud-icon cloud-icon-down"
        />
      </div>
      <ul
        class="group-sub collapse"
      >
        <li
          class="item"
          role="menuitem"
          style="text-indent:30px"
        >
          <span
            style="color:red;font-weight:bold"
          >
            朝阳区
          </span>
        </li>
        <li
          class="item"
          role="menuitem"
          style="text-indent:30px"
        >
          海淀区
        </li>
      </ul>
    </li>
    <li>
      <div
        class="group-title"
        style="text-indent:20px"
      >
        上海
        <i
          class="indicator cloud-icon cloud-icon-down"
        />
      </div>
      <ul
        class="group-sub collapse"
      >
        <li
          class="item"
          role="menuitem"
          style="text-indent:30px"
        >
          浦东区
        </li>
        <li
          class="item"
          role="menuitem"
          style="text-indent:30px"
        >
          南京路
        </li>
      </ul>
    </li>
  </ul>
</section>
`;
